triplea-game/triplea
# Fix out-of-date map check (#1522)

This PR fixes the broken out-of-date map check reported in issue #1522.

The out-of-date map check was using the wrong file name when looking for a map's download properties.  It only considered the property file `<mapName>.zip.properties`, which does not exist for map's downloaded from GitHub.  When the map download properties cannot be found, it is assumed the map is up-to-date.

Since migrating the maps to GitHub, `ResourceLoader` considers the following map zip file candidates when attempting to load a map:

1. `<mapName>.zip`
1. `<normalizedMapName>-master.zip`
1. `<normalizedMapName>.zip`

The fix for this issue was to modify the out-of-date map check to also consider all these map zip file names when attempting to locate the corresponding download properties file (the same name with `.properties` appended).

#### Functional changes
* The out-of-date map check in `MapDownloadController` now considers all possible map zip file candidates (see above) when looking for the corresponding download properties file.  Previously, it only considered candidate (1).

#### Functional issues for review
None.

#### Refactoring changes
* I extracted a new interface and several new methods in `MapDownloadController` in order to write unit tests for the out-of-date map check functionality.
* I modified `ResourceLoader#getPaths()` to use the new `ResourceLoader#getMapZipFileCandidates()` method in order to be DRY.

#### Refactoring issues for review
* The changes to `ResourceLoader#getPaths()` were not necessary to implement the fix in this PR.  As I said above, I simply did it to avoid duplication in map path candidates (which was essentially the root cause of this issue).  Please advise if that hunk should be reverted.
* The out-of-date map check has only ever considered maps whose shape is a zip file; maps with a directory shape are not supported by this feature.  Some of the names I used when refactoring reflect this limitation to make it explicitly clear.

#### Testing
I first confirmed I could reproduce the issue on `master` (1.9.0.0.4208).  I performed the following steps:

* Delete the `TRIPLEA_LAST_CHECK_FOR_MAP_UPDATES` key from the user preferences to force an out-of-date map check.
* Locally modify my `tutorial-master.zip.properties` and `the_pact_of_steel-master.zip.properties` files and change the `map.version` keys to `0.9` (they previously had the value `1.0`).  (I have a total of six maps installed, so there are still four maps present that are up-to-date.)
* Run TripleA.

After waiting for about a minute, I was not prompted to download the updated maps.

I then repeated the steps on this branch.  I confirmed I was prompted to download both updated maps.  I answered **Yes** to the prompt, downloaded the two maps in question, and exited TripleA.

I deleted the `TRIPLEA_LAST_CHECK_FOR_MAP_UPDATES` key once again and restarted TripleA.  I waited for about a minute to confirm I was not prompted to download any maps.

I also started a few games from maps whose shape is a zip file to confirm the changes in `ResourceLoader#getPaths()` did not break map loading.
